File: apps/android/native/core/src/main/kotlin/org/indiafoss/companion/core/ScheduleFreshness.kt
package org.indiafoss.companion.core

import java.time.OffsetDateTime

/**
 * Honest refresh state for the programme (#191), the Kotlin twin of
 * `apps/web/src/lib/schedule-freshness.ts`. Three clocks are kept apart
 * because conflating them is what makes an old schedule look current:
 *
 *  - `sourceMetadata.sourceUpdatedAt` — when the programme was last imported
 *    from the organisers.
 *  - `lastCheckedAt` — when this device last reached the manifest. It says
 *    nothing about how old the data is.
 *  - `sourceMetadata.scheduleStatus` — whether the organisers still call the
 *    programme a draft. Absent in older bundles, and absence is never reported
 *    as confirmation.
 *
 * A bundle read out of the APK's assets rather than downloaded is as old as
 * the release, so [SeedSource.SEED] is reported too.
 *
 * Every value comes from the published bundle. Nothing here invents a
 * timestamp or a sync mechanism.
 */
object ScheduleFreshness {
    const val CURRENT_MAX_MS = 2 * 24 * 60 * 60_000L
    const val STALE_MIN_MS = 7 * 24 * 60 * 60_000L
    const val CHECK_OVERDUE_MS = 24 * 60 * 60_000L

    /** How old the imported programme data is. */
    enum class Age { UNKNOWN, CURRENT, AGEING, STALE }

    /** Where the rendered bundle came from; mirrors the repository's own distinction. */
    enum class SeedSource { REFRESHED, SEED, NONE }

    data class State(
        val provisional: Boolean,
        val statusLine: String?,
        val importedAtMs: Long?,
        val importedLine: String,
        val age: Age,
        val checkOverdue: Boolean,
        val checkedLine: String,
        val sourceLine: String?,
    )

    /** The bundle's own UTC offset, taken from its start instant (e.g. `+05:30`). */
    fun offsetOf(start: String?): String? {
        val match = Regex("(Z|[+-]\\d{2}:\\d{2})$").find(start.orEmpty()) ?: return null
        return if (match.value == "Z") "+00:00" else match.value
    }

    /**
     * Upstream sends a naive `YYYY-MM-DD HH:MM:SS[.ffffff]` timestamp with no
     * zone. It is anchored to the event's own offset rather than to the phone's
     * zone, which would be wrong by hours for anyone travelling.
     */
    fun parseSourceTimestamp(raw: String?, offset: String?): Long? {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        val zoned = when {
            Regex("(Z|[+-]\\d{2}:\\d{2})$").containsMatchIn(trimmed) -> trimmed.replace(' ', 'T')
            offset != null -> trimmed.replace(' ', 'T') + offset
            else -> return null
        }
        return runCatching { OffsetDateTime.parse(zoned).toInstant().toEpochMilli() }.getOrNull()
    }

    /** Coarse on purpose, so it never looks more precise than it is. */
    fun describeElapsed(elapsedMs: Long): String {
        if (elapsedMs < 60 * 60_000L) return "less than an hour ago"
        val hours = elapsedMs / (60 * 60_000L)
        if (hours < 24) return "$hours hour${if (hours == 1L) "" else "s"} ago"
        val days = hours / 24
        return "$days day${if (days == 1L) "" else "s"} ago"
    }

    fun describe(
        bundle: EventBundle?,
        lastCheckedAt: Long?,
        source: SeedSource,
        now: Long = System.currentTimeMillis(),
    ): State {
        val metadata = bundle?.sourceMetadata
        val importedAtMs = parseSourceTimestamp(metadata?.sourceUpdatedAt, offsetOf(bundle?.start))
        val age = when {
            importedAtMs == null -> Age.UNKNOWN
            now - importedAtMs < CURRENT_MAX_MS -> Age.CURRENT
            now - importedAtMs < STALE_MIN_MS -> Age.AGEING
            else -> Age.STALE
        }

        val provisional = metadata?.scheduleStatus == "draft"
        val statusLine = when (metadata?.scheduleStatus) {
            null, "" -> null
            "draft" -> "Provisional: the organisers have not marked this programme final. Times and rooms may still change."
            else -> "The organisers have marked this programme final."
        }

        val importedLine = if (importedAtMs == null) {
            "This schedule does not record when it was imported."
        } else {
            "Programme data imported from the organisers ${describeElapsed(now - importedAtMs)}." +
                if (age == Age.STALE) " No newer import has been published, so treat these times as out of date." else ""
        }

        val overdue = lastCheckedAt == null || now - lastCheckedAt >= CHECK_OVERDUE_MS
        val checkedLine = if (lastCheckedAt == null) {
            "This device has not checked for a newer programme yet, so a newer one may exist."
        } else {
            "This device last reached the organisers ${describeElapsed(now - lastCheckedAt)}. " +
                "That is when it looked, not how old the programme is."
        }

        val sourceLine = when (source) {
            SeedSource.SEED -> "Showing the copy built into this app release; no update has been downloaded yet."
            SeedSource.NONE -> "No schedule is stored on this device."
            SeedSource.REFRESHED -> null
        }

        return State(
            provisional = provisional,
            statusLine = statusLine,
            importedAtMs = importedAtMs,
            importedLine = importedLine,
            age = age,
            checkOverdue = overdue,
            checkedLine = checkedLine,
            sourceLine = sourceLine,
        )
    }
}

File: apps/web/src/lib/schedule-freshness.ts
import type { EventBundle } from '@indiafoss/model';

/**
 * Honest refresh state for the programme (#191).
 *
 * The Settings card used to show one number: when this device last reached the
 * server. A successful check therefore read as "your schedule is current", even
 * when the newest published revision was itself an old, still-provisional
 * import. Three different clocks matter and they are kept apart here:
 *
 *  - `sourceMetadata.sourceUpdatedAt` — when the upstream programme last
 *    changed, i.e. when the data in this bundle was imported from FOSS United.
 *  - `lastCheckedAt` — when *this device* last successfully reached the
 *    manifest. It says nothing about how old the data is.
 *  - `sourceMetadata.scheduleStatus` — whether the organisers call this
 *    programme a draft. Absent in older bundles, and absence is not a claim
 *    that anything has been confirmed.
 *
 * Nothing here invents a sync mechanism or a timestamp: every field comes from
 * the bundle that is already published.
 */

/** How old the imported programme data is, in plain buckets. */
export type ScheduleAge = 'unknown' | 'current' | 'ageing' | 'stale';

/** Under two days old counts as current. */
export const CURRENT_MAX_MS = 2 * 24 * 60 * 60_000;
/** Over a week old is called stale outright. */
export const STALE_MIN_MS = 7 * 24 * 60 * 60_000;
/** A device that has not reached the server in a day is told so. */
export const CHECK_OVERDUE_MS = 24 * 60 * 60_000;

export interface ScheduleFreshnessInput {
  /** The bundle actually being rendered, or null before one is loaded. */
  bundle: Pick<EventBundle, 'start' | 'sourceMetadata'> | null;
  /** Epoch ms of this device's last successful manifest check, if any. */
  lastCheckedAt: number | null;
  /** Wall-clock now. Deliberately not the day simulator: staleness is a real-world fact. */
  now: number;
}

export interface ScheduleFreshness {
  /** True only when the bundle says so. Absence is reported as unknown, never as confirmed. */
  provisional: boolean;
  /** One line about publication status, or null when the bundle does not say. */
  statusLine: string | null;
  /** Normalised ISO instant of the upstream import, or null when unusable. */
  importedAtIso: string | null;
  /** One line naming when the programme data was imported. */
  importedLine: string;
  age: ScheduleAge;
  /** Whether this device is overdue for a check; not a statement about the data. */
  checkOverdue: boolean;
  /** One line about this device's own last check. */
  checkedLine: string;
}

/** The bundle's own UTC offset, taken from its start instant (e.g. `+05:30`). */
export function bundleOffset(start: string | undefined): string | null {
  const match = /(Z|[+-]\d{2}:\d{2})$/.exec(start ?? '');
  if (!match) return null;
  return match[1] === 'Z' ? '+00:00' : match[1]!;
}

/**
 * Upstream sends a naive `YYYY-MM-DD HH:MM:SS[.ffffff]` timestamp with no zone.
 * `Date.parse` treats that inconsistently across browsers, so it is anchored to
 * the event's own offset rather than to whatever zone the phone is in. Already
 * zoned values are passed through untouched.
 */
export function parseSourceTimestamp(
  raw: string | undefined,
  offset: string | null,
): string | null {
  if (!raw) return null;
  const trimmed = raw.trim();
  const zoned = /(Z|[+-]\d{2}:\d{2})$/.test(trimmed)
    ? trimmed.replace(' ', 'T')
    : offset
      ? `${trimmed.replace(' ', 'T')}${offset}`
      : null;
  if (!zoned) return null;
  const parsed = Date.parse(zoned);
  return Number.isFinite(parsed) ? new Date(parsed).toISOString() : null;
}

function ageOf(importedAtIso: string | null, now: number): ScheduleAge {
  if (!importedAtIso) return 'unknown';
  const elapsed = now - Date.parse(importedAtIso);
  if (!Number.isFinite(elapsed) || elapsed < 0) return 'current';
  if (elapsed < CURRENT_MAX_MS) return 'current';
  return elapsed < STALE_MIN_MS ? 'ageing' : 'stale';
}

/** "3 days", "5 hours", "just now" — coarse on purpose, so it never looks precise. */
export function describeElapsed(elapsedMs: number): string {
  if (elapsedMs < 60 * 60_000) return 'less than an hour ago';
  const hours = Math.floor(elapsedMs / (60 * 60_000));
  if (hours < 24) return `${hours} hour${hours === 1 ? '' : 's'} ago`;
  const days = Math.floor(hours / 24);
  return `${days} day${days === 1 ? '' : 's'} ago`;
}

export function describeScheduleFreshness(input: ScheduleFreshnessInput): ScheduleFreshness {
  const { bundle, lastCheckedAt, now } = input;
  const metadata = bundle?.sourceMetadata;
  const importedAtIso = parseSourceTimestamp(
    metadata?.sourceUpdatedAt,
    bundleOffset(bundle?.start),
  );
  const age = ageOf(importedAtIso, now);

  const provisional = metadata?.scheduleStatus === 'draft';
  const statusLine = !metadata?.scheduleStatus
    ? null
    : provisional
      ? 'Provisional: the organisers have not marked this programme final. Times and rooms may still change.'
      : 'The organisers have marked this programme final.';

  const importedLine = !importedAtIso
    ? 'This schedule does not record when it was imported.'
    : `Programme data imported from the organisers ${describeElapsed(now - Date.parse(importedAtIso))}.`;

  const overdue = lastCheckedAt === null || now - lastCheckedAt >= CHECK_OVERDUE_MS;
  const checkedLine =
    lastCheckedAt === null
      ? 'This device has not checked for a newer programme yet, so a newer one may exist.'
      : `This device last reached the organisers ${describeElapsed(now - lastCheckedAt)}. That is when it looked, not how old the programme is.`;

  return {
    provisional,
    statusLine,
    importedAtIso,
    importedLine,
    age,
    checkOverdue: overdue,
    checkedLine,
  };
}
